### DISCUSSION TITLE
HEAT V4 differentiation of metadata & properties

### DISCUSSION THREAD

**Discussion by MarkStega** [wixtoolset/issues#7304] [Questions]
I am not understanding the difference between using metadata & properties to control harvesting

As an example the following does not suppress the indicated elements that I want suppressed:
```
    <HarvestDirectory Include="$(SolutionDir)FileCleanupService.2012\bin"
					  ComponentGroupName="FileCleanupService"
					  HarvestDirectorySuppressCom="true"
					  HarvestDirectorySuppressFragments="true"
					  HarvestDirectorySuppressRegistry="true"
					  HarvestDirectorySuppressRootDirectory="true"
					  HarvestDirectoryTransforms="$(ProjectDir)Filter.xslt" />
```
but the following works:
```
    <HarvestDirectory Include="$(SolutionDir)FileCleanupService.2012\bin\" 
					  HarvestDirectoryTransforms="$(ProjectDir)Filter.xslt">
      <ComponentGroupName>FileCleanupService</ComponentGroupName>
      <DirectoryRefId>APPLICATIONROOTDIRECTORY1</DirectoryRefId>
      <SuppressCom>true</SuppressCom>
      <SuppressRegistry>true</SuppressRegistry>
      <SuppressRootDirectory>true</SuppressRootDirectory>
    </HarvestDirectory>
```
Am I just not applying the properties correctly? The documentation for the properties is a bit confusing with respect to how to use them...

**Comment by robmen** [comment]
Those are two different ways to represent the same thing: MSBuild metadata [as properties](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-element-msbuild?view=vs-2022) or [as attributes](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-element-msbuild?view=vs-2022#specify-metadata-as-attributes).

However, you are using different names for the Metadata in your two examples. That is probably why you see different behavior.

**Comment by rseanhall** [comment]
`HarvestDirectorySuppressRegistry` is an example of a property. `SuppressCom` is an example of item metadata. Properties are declared as an element inside of the `PropertyGroup` element, not as metadata. For that example, this means that they are applied to all `HarvestDirectory` elements in the project.

**Reply by MarkStega** [reply]
@robmen @rseanhall 

Just to verify that I have this correct:

Any of the metadata can be expressed either as an attribute or a property, ie:
```
		<HarvestDirectory Include="$(SolutionDir)CommunicationService.2012\bin\">
			<ComponentGroupName>CommunicationService</ComponentGroupName>
		</HarvestDirectory>

and

		<HarvestDirectory Include="$(SolutionDir)CommunicationService.2012\bin\"
			                      ComponentGroupName="CommunicationService"
		</HarvestDirectory>
```
are equivalent uses of metadata.

Using a property is limited to a PropertyGroup and becomes a global setting so
```
<PropertyGroup>
    <HarvestDirectorySuppressCom>true </HarvestDirectorySuppressCom>
</PropertyGroup>
```
sets the suppression of COM elements for all <HarvestDirectory> elements.

1) For something such as suppress COM which is defined as both metadata & property what takes precedent if both are specified? (I would hope that the metadata overrides the property.)

2) For anything settable as a property that doesn't have a corresponding metadata definition I presume it would be a code change to support the metadata? An example would be that a property exists for "SuppressFragments" but there is no metadata to control that so it would be an all or none decision.
